### ISSUE TITLE
document range types syntax and semantics

### ISSUE THREAD

**Issue by RossBoylan** [JuliaLang/julia#16393]
Range's and the : notation are mentioned in passing in several spots in the documentation, but it doesn't seem there's any thorough presentation about them.

Compare Python's documentation Slicings in 6.3.3 of https://docs.python.org/3/reference/expressions.html#primaries and the discussion of : in sequences at https://docs.python.org/3/library/stdtypes.html#sequence-types-list-tuple-range and https://docs.python.org/3/library/functions.html#func-range

Also, x:y:z has step y in julia but z in python.  Maybe julia should match python?  The change would be bad for old code, but might prevent errors for people coming from python.

Possible topics:
- Behavior if types of x, y, z differ.
- If y<x, x:y generates nothing (its use as a special marker for a missing match is discussed in the curent docs).  Note that in R, but not Python, 3:1 generates a descending sequence.
- But if y<0 then x:y:z generates a descending range if z<x.
- Precedence of : relative to other operators.
- Numerical issues when z-x is apparently an integer multiple of y but is not.  I would expect there would be circumstances in which z would not be an element of the generated set, but I can't generate any examples!
- Circumstances under which ranges can be used in place of vector/array/tuple types, and how to convert to same.

**Comment by pao** [comment]
> Also, x:y:z has step y in julia but z in python. Maybe julia should match python? The change would be bad for old code, but might prevent errors for people coming from python.

Julia uses the MATLAB/Octave convention for range steps. So you have to pick which group of users have to adjust. Since the rest of our range syntax follows from MATLAB as well (`end`, one-based indexes (at least for normal everyday arrays)) this is unlikely to be changed.

**Comment by RossBoylan** [comment]
I agree that's a good reason to leave it as from:step:end.  You can also add R to the have to adjust group, since they have seq(from, end, step).

@tkelman thanks for the tag.  I need to find out how to add them.

The main issue is the documentation.

**Comment by nalimilan** [comment]
Would you make a pull request? Other than that, it's unlikely to happen soon.

**Comment by omus** [comment]
Ranges using `DateTime` are also undocumented. I'll note that they are rather special as the step is a different type:
```julia
julia> DateTime(2015):Dates.Day(1):DateTime(2016)
2015-01-01T00:00:00:1 day:2016-01-01T00:00:00
```

**Comment by dpsanders** [comment]
1ñ

**Comment by RossBoylan** [comment]
Why was this closed?
